src/irr/asset/COBJMeshFileLoader.cpp

```diff
@@ -355,490 +734,377 @@ asset::SAssetBundle COBJMeshFileLoader::loadAsset(io::IReadFile* _file, const as
-		mtlReader = FileSystem->createAndOpenFile(relPath + io::IFileSystem::getFileBasename(realFile));
-	if (!mtlReader)	// fail to open and read file
-	{
-		os::Printer::log("Could not open material file", realFile.c_str(), ELL_WARNING);
-		return;
-	}
+    auto vs = getDefaultAsset<ICPUSpecializedShader,IAsset::ET_SPECIALIZED_SHADER>(_hasUV ? VERT_SHADER_UV_CACHE_KEY : VERT_SHADER_NO_UV_CACHE_KEY, AssetManager);
 
-	const long filesize = mtlReader->getSize();
-	if (!filesize)
-	{
-		os::Printer::log("Skipping empty material file", realFile.c_str(), ELL_WARNING);
-		mtlReader->drop();
-		return;
-	}
+    const uint32_t presentMapsMask = (_mtl.std140PackedData.extra>>4u);//first 4 bits is illum model
+    const std::string fs_cache_key = (_hasUV ? (FRAG_SHADER_UV_CACHE_KEY + std::to_string(presentMapsMask)) : FRAG_SHADER_NO_UV_CACHE_KEY);
+    auto fs = getDefaultAsset<ICPUSpecializedShader, IAsset::ET_SPECIALIZED_SHADER>(fs_cache_key.c_str(), AssetManager);
 
-	char* buf = new char[filesize];
-	mtlReader->read((void*)buf, filesize);
-	const char* bufEnd = buf+filesize;
+    if (!fs)
+    {
+        const char* src = (_hasUV ? FRAG_SHADER_UV : FRAG_SHADER_NO_UV);
+        std::string fs_source;
+        fs_source.resize(strlen(src)+1000ull);
+        sprintf(fs_source.data(), src, genGLSLtextureBindingsStr(_mtl, _hasUV).c_str());
+        auto fs_unspec = core::make_smart_refctd_ptr<ICPUShader>(fs_source.c_str());
+        ICPUSpecializedShader::SInfo specinfo({}, nullptr, "main", ICPUSpecializedShader::ESS_FRAGMENT);
+        fs = core::make_smart_refctd_ptr<ICPUSpecializedShader>(std::move(fs_unspec), std::move(specinfo));
+        insertShaderIntoCache(fs, fs_cache_key.c_str(), AssetManager); //not calling insertShaderIntoCache through override because i want it to insert regardless of caching flags
+    }
 
-	SObjMtl* currMaterial = 0;
+    return {std::move(vs), std::move(fs)};
+}
 
-	const char* bufPtr = buf;
-	while(bufPtr != bufEnd)
-	{
-		switch(*bufPtr)
-		{
-			case 'n': // newmtl
-			{
-				// if there's an existing material, store it first
-				if ( currMaterial )
-					_ctx.Materials.push_back( currMaterial );
+auto COBJMeshFileLoader::loadImages(const char* _relDir, const CMTLPipelineMetadata::SMtl& _mtl, uint32_t _hierarchyLvl) -> images_set_t
+{
+    std::array<core::smart_refctd_ptr<ICPUImage>, CMTLPipelineMetadata::SMtl::EMP_COUNT> images;
 
-				// extract new material's name
-				char mtlNameBuf[WORD_BUFFER_LENGTH];
-				bufPtr = goAndCopyNextWord(mtlNameBuf, bufPtr, WORD_BUFFER_LENGTH, bufEnd);
+    std::string relDir = _relDir;
+    for (uint32_t i = 0u; i < images.size(); ++i)
+    {
+        SAssetLoadParams lp;
+        if (_mtl.maps[i].size())
+        {
+            auto bundle = interm_getAssetInHierarchy(AssetManager, relDir + _mtl.maps[i], lp, _hierarchyLvl);
+            if (!bundle.isEmpty())
+                images[i] = core::smart_refctd_ptr_static_cast<ICPUImage>(bundle.getContents().first[0]);
+        }
+    }
```

Comment:
shouldn't the MTL loader load the textures and create the descriptor set?

Reply:
https://github.com/buildaworldnet/IrrlichtBAW/pull/410#discussion_r367025896